# Анализ и обработка результатов

В анализе данных важно наглядное представление результатов, чтобы их мог оценить заказчик. Это бизнес, где на кону громадный трафик, серверные мощности, личное время многотысячной аудитории. И одновременно это исследование.

Как в настоящей экспериментальной науке, ответ непредсказуем. Его точность зависит от умелого владения статистическими методами и от качества исходных данных. Перед тем, как начинать считать, проверьте, грамотно ли ваши данные подготовлены.

# Группировка данных

Слово «анализ» означает разбор, рассмотрение с разных сторон. Анализ данных начинают с разделения их на группы по какому-нибудь признаку. Эта операция называется группировка данных. Она помогает изучить материал более подробно, чтобы затем перейти к поиску взаимосвязей между отдельными группами.


Группировка оправданна, если данные чётко делятся по значимому признаку, а полученные группы близки к теме задачи. Например, когда есть данные обо всех покупках в супермаркете, можно смело заниматься группировкой. Так можно установить время наплыва покупателей и решить проблему пиковых нагрузок. Или посчитать средний чек — обычно для магазинов это ключевая метрика.


Стадии группировки хорошо укладываются в словесную формулу **split-apply-combine**:

•	разделить, **split** — разбиение на группы по определённому критерию;

•	применить, **apply** — применение какого-либо метода к каждой группе в отдельности, например, подсчёт численности группы методом count() или суммирование вызовом sum();

•	объединить, **combine** — сведение результатов в новую структуру данных, в зависимости от условий разделения и выполнения метода это бывает DataFrame и Series.

На картинке изображен принцип **split-apply-combine** для таблицы с экзопланетами. Посмотрим, как вообще идут дела с поиском экзопланет. Сначала данные делят по группам, где каждая группа — это год. Потом метод **count()** подсчитывает численность каждой группы. В итоге получаем новую структуру данных с группами, где каждая содержит год и число открытых за этот год экзопланет.

![image.png](attachment:image.png)

В Рandas для группировки данных есть метод **groupby()**. Он принимает как аргумент название столбца, по которому нужно группировать. В случае с делением экзопланет по годам открытия:

**print(exoplanet.groupby('discovered'))**

**<pandas.core.groupby.DataFrameGroupBy object at 0x7fc1e1ca3400>**


Применение метода **groupby()** к объекту типа DataFrame приводит к созданию объекта особого типа — **DataFrameGroupBy**. Это сгруппированные данные. Если применить к ним какой-нибудь метод Pandas, они станут новой структурой данных типа **DataFrame** или **Series**.
Подсчитаем сгруппированные по годам экзопланеты методом **count()**:

Если нужно сравнить наблюдения по одному показателю, метод применяют к **DataFrameGroupBy** с указанием на один столбец. Нас в первую очередь интересует радиус экзопланет: мы ищем другую Землю. Давайте получим таблицу с единственным столбцом 'radius':

**exo_number = exoplanet.groupby('discovered')['radius'].count()**

**print(exo_number)**

Получили Series, где по годам открытия расписано количество экзопланет, для которых удалось установить радиус.

Посмотрим, как меняется средний радиус открытых экзопланет год от года. Для этого надо сложить радиусы планет, открытых за определённый год, и поделить на их количество (которое мы уже нашли).

Сумма радиусов считается методом **sum()**:

**exo_radius_sum = exoplanet.groupby('discovered')['radius'].sum()**


**print(exo_radius_sum)**



## Задача
Вашему вниманию предлагается датасет с обзором показателей самоубийств с 1985 по 2016 год — Сравнение социально-экономической информации с показателями самоубийств по годам и странам.

Столбцами этой таблицы являются: страна, год, пол, возрастная группа, количество самоубийств, население, уровень самоубийств, сводный ключ по стране за год, ИЧР (индекс человеческого развития) за год, ВВП за год, ВВП на душу населения, поколение (на основе среднего по возрастной группе).

1. Загрузите таблицу master, содержащую данные о самоубийствах. Выведите на экран таблицу и оцените данные, обратите внимание на названия столбцов, переименуйте их, если это необходимо.

2. Убедитесь в том, что данные прошли предподготовку.  Пропущенные и неопределённые значения выявляет метод isna(), а суммарное количество таких значений — метод sum().

3. Сгруппируйте DataFrame по столбцу country, сохраните полученный результат в переменной country_grouping.

4. Посчитайте количество случаев самоубийств по странам, которые были совершены, методом **sum()**, указав, что выбираем один столбец suicides_no.

Сохраните результат в переменной country_sum и выведите первые 30 строк этой таблицы.



In [6]:
import pandas as pd

data = pd.read_csv('master.csv')

data.columns = [
    'country', 'year', 'sex', 'age', 'suicides_no', 'population',
    'suicides_per_100k_pop', 'country_year', 'HDI_for_year',
    'gdp_for_year', 'gdp_per_capita', 'generation'
]

print("Первые строки таблицы:")
print(data.head())

missing_data = data.isna().sum()
print("\nПропущенные значения до очистки:")
print(missing_data)

data_cleaned = data.dropna()

missing_data_cleaned = data_cleaned.isna().sum()
print("\nПропущенные значения после очистки:")
print(missing_data_cleaned)

country_grouping = data_cleaned.groupby('country')

country_sum = country_grouping['suicides_no'].sum()

# Вывод первых 30 строк результата
print("\nПервые 30 строк по странам:")
print(country_sum.head(30))



Первые строки таблицы:
   country  year     sex          age  suicides_no  population  \
0  Albania  1987    male  15-24 years           21      312900   
1  Albania  1987    male  35-54 years           16      308000   
2  Albania  1987  female  15-24 years           14      289700   
3  Albania  1987    male    75+ years            1       21800   
4  Albania  1987    male  25-34 years            9      274300   

   suicides_per_100k_pop country_year  HDI_for_year   gdp_for_year  \
0                   6.71  Albania1987           NaN  2,156,624,900   
1                   5.19  Albania1987           NaN  2,156,624,900   
2                   4.83  Albania1987           NaN  2,156,624,900   
3                   4.59  Albania1987           NaN  2,156,624,900   
4                   3.28  Albania1987           NaN  2,156,624,900   

   gdp_per_capita       generation  
0             796     Generation X  
1             796           Silent  
2             796     Generation X  
3          

# Сортировка данных


Поиск необычного в группе — что среди планет, что среди меломанов — это прежде всего поиск чемпионов: объектов с выдающимися показателями по разным статьям. Как всю таблицу, так и отдельные группы изучают, сортируя строки по какому-либо столбцу.

В Pandas для этой операции есть метод **sort_values()**. У него два аргумента:

• **by = 'имя столбца'** — имя столбца, по которому нужно сортировать;

• **ascending:** по умолчанию True. Для сортировки по убыванию установите значение False.

![image.png](attachment:image.png)


Среди экзопланет интересны близкие по размерам к Земле. Есть ли такие? Отсортируем список по радиусу в порядке возрастания. Тогда в голове таблицы окажутся самые малые, на которых гравитация не прижмёт нас к полу.

**print(exoplanet.sort_values(by='radius').head(30))**

Оказывается, некоторые из уже открытых экзопланет по размерам близки не то что к Земле, но уже и к Луне! Получим список экзопланет с радиусом меньше земного. Смотрите, как логический оператор (здесь это <) задействуется в отборе элементов столбца. Дальше нам этот приём не раз понадобится.

**print(exoplanet[exoplanet['radius'] < 1])**

Но и этот список такой длинный, что изучать его лучше по частям. Экзопланеты, близкие по размерам к Земле, за последнее десятилетие открывали нередко. Можно изучать список открытых за каждый год. Например, для 2014 года (вновь обратите внимание на работу логического оператора, теперь это ==):

**print(exoplanet[exoplanet['discovered'] == 2014])**

А чтобы не тратить время на лишнее, поставим оба условия сразу. Для этого в Pandas есть логический оператор &, подобный оператору and языка Python. Напомним, его смысл на русском языке можно передать словами «и ещё»:

**exo_small_14 = exoplanet[ (exoplanet['radius']<1) & (exoplanet['discovered']==2014)]**

**print(exo_small_14)**

Отсортируем результат в порядке убывания радиуса.

**print(exo_small_14.sort_values(by = 'radius', ascending = False))**





## Задача
1. Выделим пятерки стран с самым низким и самым высоким уровнем самоубийств в мире.

In [7]:
top_5_highest_suicides = country_sum.nlargest(5)
top_5_lowest_suicides = country_sum.nsmallest(5)

print("Топ-5 стран с самым высоким количеством самоубийств:")
print(top_5_highest_suicides)

print("\nТоп-5 стран с самым низким количеством самоубийств:")
print(top_5_lowest_suicides)



Топ-5 стран с самым высоким количеством самоубийств:
country
United States    355963
Japan            259729
France           106296
Germany           98542
Ukraine           85362
Name: suicides_no, dtype: int64

Топ-5 стран с самым низким количеством самоубийств:
country
Antigua and Barbuda     0
Grenada                 3
Maldives                4
Oman                   20
Seychelles             32
Name: suicides_no, dtype: int64


In [ ]:
#Youre code is here


2. Исследуйте распределение количества суидидов по гендерному признаку. Найдите лидеров и антилидеров по количеству самоубийств среди женщин. Повторяется ли картина по сравнению с предыдущим исследованием?

In [8]:
# Сгруппируем данные по полу и странам для анализа самоубийств среди мужчин и женщин
gender_grouping = data_cleaned.groupby(['country', 'sex'])['suicides_no'].sum().unstack()

# Найдем лидеров и антилидеров по количеству самоубийств среди женщин
top_5_female_suicides = gender_grouping['female'].nlargest(5)
bottom_5_female_suicides = gender_grouping['female'].nsmallest(5)

# Найдем лидеров и антилидеров по количеству самоубийств среди мужчин
top_5_male_suicides = gender_grouping['male'].nlargest(5)
bottom_5_male_suicides = gender_grouping['male'].nsmallest(5)

# Выводим результаты
print("Топ-5 стран с самым высоким количеством самоубийств среди женщин:")
print(top_5_female_suicides)

print("\nТоп-5 стран с самым низким количеством самоубийств среди женщин:")
print(bottom_5_female_suicides)

print("\nТоп-5 стран с самым высоким количеством самоубийств среди мужчин:")
print(top_5_male_suicides)

print("\nТоп-5 стран с самым низким количеством самоубийств среди мужчин:")
print(bottom_5_male_suicides)


Топ-5 стран с самым высоким количеством самоубийств среди женщин:
country
Japan            80683
United States    75033
France           28168
Germany          26623
Brazil           17736
Name: female, dtype: int64

Топ-5 стран с самым низким количеством самоубийств среди женщин:
country
Antigua and Barbuda    0
Grenada                0
Maldives               0
Oman                   0
Seychelles             3
Name: female, dtype: int64

Топ-5 стран с самым высоким количеством самоубийств среди мужчин:
country
United States    280930
Japan            179046
France            78128
Germany           71919
Ukraine           69424
Name: male, dtype: int64

Топ-5 стран с самым низким количеством самоубийств среди мужчин:
country
Antigua and Barbuda     0
Grenada                 3
Maldives                4
Oman                   20
Bahamas                29
Name: male, dtype: int64


# Описательная статистика

Чтобы представить себе среднестатистического самоубийцу, понадобится описательная статистика.

Из её показателей для количественного описания данных в нашей задаче нужны четыре меры — максимум, минимум, медиана и среднее.

Наибольшее и наименьшее обычно вычисляют только по одному признаку. Например, можно получить минимальное и максимальное значение уровня самоубийств (количество самоубийств на 100 тыс.населения) (столбец 'suicides/100k pop'). Для поиска максимума вызывают метод max(), примененный к этому столбцу:

print(master['suicides/100k pop'].max())


In [11]:
print(data['suicides_per_100k_pop'].max())

224.97


Самый высокий показатель оказался 224.97 самоубийств на 100 тыс.населения. Интересно, в какой стране такая ситуация.

Запросим из master строку с максимальным значением, прибегнув к логической индексации с условием

**master['suicides/100k pop'] == master['suicides/100k pop'].max()**

In [14]:
# Найдем максимальное значение в столбце 'suicides_per_100k_pop'
max_suicide_rate = data_cleaned['suicides_per_100k_pop'].max()

# Используем логическую индексацию для поиска строки с этим значением
max_suicide_row = data_cleaned[data_cleaned['suicides_per_100k_pop'] == max_suicide_rate]

# Выводим строку с максимальным уровнем самоубийств
print(max_suicide_row)


        country  year   sex        age  suicides_no  population  \
24332  Suriname  2012  male  75+ years           10        5346   

       suicides_per_100k_pop  country_year  HDI_for_year   gdp_for_year  \
24332                 187.06  Suriname2012         0.711  4,980,000,000   

       gdp_per_capita generation  
24332           10227     Silent  


Минимальное значение — ищут методом **min()**. Нам интересно установить случаи, когда уровень суицидов минимальный, но не нулевой.

Вот как мы создаём выборку  и находим в ней минимальное значение:

**master_drop_null = master[master['suicides/100k pop'] != 0]**

**print(master_drop_null['suicides/100k pop'].min())**


In [13]:
master_drop_null = data[data['suicides_per_100k_pop'] != 0]
print(master_drop_null['suicides_per_100k_pop'].min())

0.02


Получим название самой "счастливой" страны. Как и при поиске максимума, воспользуемся условием и логической индексацией

In [15]:
# Найдем минимальное ненулевое значение в столбце 'suicides_per_100k_pop'
min_suicide_rate = data_cleaned[data_cleaned['suicides_per_100k_pop'] > 0]['suicides_per_100k_pop'].min()

# Используем логическую индексацию для поиска строки с этим значением
min_suicide_row = data_cleaned[data_cleaned['suicides_per_100k_pop'] == min_suicide_rate]

# Выводим строку с минимальным ненулевым уровнем самоубийств
print(min_suicide_row)


            country  year     sex         age  suicides_no  population  \
23347  South Africa  2000  female  5-14 years            1     5079980   
23479  South Africa  2011    male  5-14 years            1     5198162   

       suicides_per_100k_pop      country_year  HDI_for_year     gdp_for_year  \
23347                   0.02  South Africa2000         0.632  136,361,854,808   
23479                   0.02  South Africa2011         0.651  416,878,162,441   

       gdp_per_capita    generation  
23347            3361    Millenials  
23479            8906  Generation Z  


Результат вывода — две страны, которые соответствуют минимальному значению.
На основе полученных данных можно сделать вывод, что число совершенных самоубийств на 100 тыс. населения  находится в диапазоне от 0,02 до 224.97 секунд, не включая пропущенные.

Это знание пригодится, чтобы разобраться с медианой и средним арифметическим.

Среднее и медиана оценивают значения в центре выборки. Если тех строк, где число с большим количеством самоубийц, столько же, сколько тех, где это число мало — среднее подойдет. Но когда есть оторванные от основной массы лидеры, их результаты сильно смещают значение среднего вверх. Вот почему оценивать в этом случае лучше медианой.

![image.png](attachment:image.png)

В геометрии медиана делит фигуру на две равные по площади. В статистике она делит выборку пополам: в одной половине значения меньше медианного, в другой больше. Логично, что для определения медианы список обязательно должен быть отсортирован — либо по возрастанию, либо по убыванию.
Когда количество значений нечётное, медиана будет равна тому значению, которое оказалось ровно посередине отсортированного набора. Если же количество данных чётное, то медиана рассчитывается как среднее арифметическое двух соседних чисел в середине набора.



В Pandas есть метод **median()**, который считает медиану. По аналогии с **min()** и **max()** его можно применять ко всей таблице, к отдельному столбцу или к сгруппированным данным. Теперь применим медиану ко всем значениям времени прослушивания в нашей таблице, исключив нулевые:



In [16]:
# Применим медиану к столбцу 'suicides_per_100k_pop', исключив нулевые значения
median_suicide_rate = data_cleaned[data_cleaned['suicides_per_100k_pop'] > 0]['suicides_per_100k_pop'].median()

# Выводим медиану
print(median_suicide_rate)


7.64


Чтобы убедиться, что лидеры действительно смещают средний показатель вверх, найдём среднее арифметическое всех этих значений методом mean():

In [17]:
# Рассчитаем среднее арифметическое для столбца 'suicides_per_100k_pop', исключая нулевые значения
mean_suicide_rate = data_cleaned[data_cleaned['suicides_per_100k_pop'] > 0]['suicides_per_100k_pop'].mean()

# Выводим среднее значение
print(mean_suicide_rate)


13.917101429166088
